src/aws-cpp-sdk-core/include/smithy/identity/signer/built-in/SigV4aSigner.h

```diff
@@ -0,0 +1,202 @@
+#include <aws/crt/auth/Credentials.h>
+
+#include <aws/core/http/HttpRequest.h>
+#include <aws/core/auth/signer/AWSAuthSignerHelper.h>
+#include <aws/crt/http/HttpConnection.h>
+#include <aws/crt/http/HttpRequestResponse.h>
+
+
+namespace smithy {
+    static const char* UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD";
+    static const char* EMPTY_STRING_SHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+    static const char v4AsymmetricLogTag[] = "AWSAuthSymmetricV4Signer";
+    static const char* USER_AGENT = "user-agent";
+
+
+    /**
+     * A smithy SigV4 signer wrapper on top of legacy SDK SigV4 signer
+     * TODO: refactor into own signer using smithy design
+     */
+    class AwsSigV4aSigner : public AwsSignerBase<AwsCredentialIdentityBase> {
+        
+    public:
+        using SigV4aAuthSchemeParameters = smithy::DefaultAuthSchemeResolverParameters;
+        explicit AwsSigV4aSigner(const SigV4aAuthSchemeParameters& parameters)
+            :  m_parameters{parameters}
+        {
+        }
+
+        SigningFutureOutcome sign(std::shared_ptr<HttpRequest> httpRequest, const AwsCredentialIdentityBase& identity, SigningProperties properties) override
+        {
+            
+            auto signPayloadIt = properties.find("SignPayload");
+            bool signPayload = signPayloadIt != properties.end() ? signPayloadIt->second.get<Aws::String>() == "true" : false;
+
+            assert(httpRequest);
+            assert(identity.sessionToken().has_value());
+            assert(identity.expiration().has_value());
+
+            auto &request = *httpRequest;
+           
+            auto crtCredentials = Aws::MakeShared<Aws::Crt::Auth::Credentials>(v4AsymmetricLogTag,
+                Aws::Crt::ByteCursorFromCString(identity.accessKeyId().c_str()),
+                Aws::Crt::ByteCursorFromCString(identity.secretAccessKey().c_str()),
+                Aws::Crt::ByteCursorFromCString((*identity.sessionToken()).c_str()),
+                (*identity.expiration()).Seconds());
+
+            Aws::Crt::Auth::AwsSigningConfig awsSigningConfig;
+            
+            bool success = createAwsSigningConfig(crtCredentials, request, awsSigningConfig, signPayload);
+
+            if(!success)
+            {
+                AWS_LOGSTREAM_ERROR(v4AsymmetricLogTag, "Failed to get Auth configureation");
+
+                return SigningError(Aws::Client::CoreErrors::MEMORY_ALLOCATION, "", "Failed to get Auth configureation", false);
+            }
+
+            std::shared_ptr<Aws::Crt::Http::HttpRequest> crtHttpRequest = request.ToCrtHttpRequest();
+
+            auto sigv4HttpRequestSigner = Aws::MakeShared<Aws::Crt::Auth::Sigv4HttpRequestSigner>(v4AsymmetricLogTag);
+            
+            sigv4HttpRequestSigner->SignRequest(crtHttpRequest, awsSigningConfig,
+                [&request, &success, this](const std::shared_ptr<Aws::Crt::Http::HttpRequest>& signedCrtHttpRequest, int errorCode) {
```

Comment:
If this is the way it was before, we had a bug, whoopsies.

we're going to need synchronization here if you plan on returning based on checking "success" per the [documentation of CRT](https://github.com/awslabs/aws-crt-cpp/blob/52dea442dd7a495c3c691268d99eb5f5765c7cee/include/aws/crt/auth/Signing.h#L78C1-L86C42)

```
/**
 * Signs an http request based on the signing implementation and supplied configuration
 * @param request http request to sign
 * @param config base signing configuration.  Actual type should match the configuration expected
 * by the signer implementation
 * @param completionCallback completion function to invoke when signing has completed or failed
 * @return true if the signing process was kicked off, false if there was a synchronous failure.
 */
virtual bool SignRequest(
```

so we return true if the signing process was kicked off, not that _it finished_

we need to add checks for:
* that we only attempt to read success after we know the callback has run.
* if Sign encountered a synchronous failure.

Reply:
checking

Reply:
I am not sure why "so we return true if the signing process was kicked off, not that it finished" is true because the function is a sync one and I see that we do return the status on completion. Happy to discuss if I am wrong

Reply:
sigv4HttpRequestSigner->SignRequest takes a callback which gets invoked at the end of the sync workflow and updates success variable , after which we return